File: SlackOff/Danmaku/DanmakuSettings.swift
import Observation
import SwiftUI

/// 弹幕样式：颜色 / 速度 / 显示区域 / 透明度 / 字号。
/// 所有选项都会持久化到 UserDefaults，下次启动/进入播放页自动恢复。
@Observable
final class DanmakuSettings {
    private static let storageKey = "com.slackoff.danmaku-settings"

    var enabled = true { didSet { save() } }
    var opacity = 1.0 { didSet { save() } }
    var speed = 1.0 { didSet { save() } }
    var region = 1.0 { didSet { save() } }
    var fontSize = 16.0 { didSet { save() } }
    var colorPreset: DanmakuColorPreset = .original { didSet { save() } }
    /// 防重叠：让同时出现的弹幕尽量落到不同轨道。
    var antiOverlap = true { didSet { save() } }

    init() {
        load()
    }

    private func save() {
        let dict: [String: Any] = [
            "enabled": enabled,
            "opacity": opacity,
            "speed": speed,
            "region": region,
            "fontSize": fontSize,
            "colorPreset": colorPreset.rawValue,
            "antiOverlap": antiOverlap
        ]
        UserDefaults.standard.set(dict, forKey: Self.storageKey)
    }

    private func load() {
        guard let dict = UserDefaults.standard.dictionary(forKey: Self.storageKey) else { return }
        enabled = dict["enabled"] as? Bool ?? true
        opacity = dict["opacity"] as? Double ?? 1.0
        speed = dict["speed"] as? Double ?? 1.0
        region = dict["region"] as? Double ?? 1.0
        fontSize = dict["fontSize"] as? Double ?? 16.0
        colorPreset = (dict["colorPreset"] as? String).flatMap(DanmakuColorPreset.init(rawValue:)) ?? .original
        antiOverlap = dict["antiOverlap"] as? Bool ?? true
    }
}

/// 弹幕颜色预设（`.original` 表示沿用每条弹幕自身的颜色）。
enum DanmakuColorPreset: String, CaseIterable, Identifiable {
    case original, white, red, yellow, green, cyan, pink, purple

    var id: String { rawValue }

    var label: String {
        switch self {
        case .original: return "默认"
        case .white: return "白"
        case .red: return "红"
        case .yellow: return "黄"
        case .green: return "绿"
        case .cyan: return "青"
        case .pink: return "粉"
        case .purple: return "紫"
        }
    }

    /// 弹幕实际显示用的颜色；`.original` 返回 nil，表示不强覆盖。
    var override: Color? {
        self == .original ? nil : tint
    }

    var tint: Color {
        switch self {
        case .original: return .white
        case .white: return .white
        case .red: return .red
        case .yellow: return .yellow
        case .green: return .green
        case .cyan: return .cyan
        case .pink: return .pink
        case .purple: return .purple
        }
    }
}

File: android/app/src/main/java/com/slackoff/app/danmaku/DanmakuSettings.kt
package com.slackoff.app.danmaku

import android.content.Context
import android.content.SharedPreferences
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.graphics.Color

enum class DanmakuColorPreset(val label: String, val tint: Color) {
    ORIGINAL("默认", Color.White),
    WHITE("白", Color.White),
    RED("红", Color.Red),
    YELLOW("黄", Color.Yellow),
    GREEN("绿", Color.Green),
    CYAN("青", Color.Cyan),
    PINK("粉", Color(0xFFFFC0CB)),
    PURPLE("紫", Color(0xFF800080));

    val override: Color?
        get() = if (this == ORIGINAL) null else tint
}

class DanmakuSettings {
    companion object {
        private const val PREFS_NAME = "com.slackoff.danmaku-settings"
    }

    // 用 Compose State 包装，设置变化时 UI 立即刷新（对齐 iOS 的 @Observable）
    var enabled: Boolean by mutableStateOf(true)
    var opacity: Float by mutableStateOf(1.0f)
    var speed: Float by mutableStateOf(1.0f)
    var region: Float by mutableStateOf(1.0f)
    var fontSize: Float by mutableStateOf(16.0f)
    var colorPreset: DanmakuColorPreset by mutableStateOf(DanmakuColorPreset.ORIGINAL)
    var antiOverlap: Boolean by mutableStateOf(true)

    private var prefs: SharedPreferences? = null

    fun load(context: Context) {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        val p = prefs!!
        enabled = p.getBoolean("enabled", true)
        opacity = p.getFloat("opacity", 1.0f)
        speed = p.getFloat("speed", 1.0f)
        region = p.getFloat("region", 1.0f)
        fontSize = p.getFloat("fontSize", 16.0f)
        colorPreset = try {
            DanmakuColorPreset.valueOf(p.getString("colorPreset", "ORIGINAL") ?: "ORIGINAL")
        } catch (_: Exception) {
            DanmakuColorPreset.ORIGINAL
        }
        antiOverlap = p.getBoolean("antiOverlap", true)
    }

    fun save() {
        val p = prefs ?: return
        p.edit()
            .putBoolean("enabled", enabled)
            .putFloat("opacity", opacity)
            .putFloat("speed", speed)
            .putFloat("region", region)
            .putFloat("fontSize", fontSize)
            .putString("colorPreset", colorPreset.name)
            .putBoolean("antiOverlap", antiOverlap)
            .apply()
    }
}